luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+	local areaOfEffect = weaponDef.damageAreaOfEffect
+	local interceptedByShieldType = weaponDef.interceptedByShieldType
+
+	if areaOfEffect > 11 and not weaponDef.customParams.shield_aoe_penetration and weaponDef.interceptedByShieldType == 1 then -- 11 because the the benchmark cortex sumo has a AOE of 12
+		AOEWeaponDefIDs[weaponDefID] = true
+	end
+
+	if weaponDef.customParams.beamtime_damage_reduction_multiplier then
+		local base = weaponDef.customParams.shield_damage
+		local multiplier = weaponDef.customParams.beamtime_damage_reduction_multiplier
+		local damage = math.max(base * multiplier)
+		originalShieldDamages[weaponDefID] = math.floor(damage)
+	else
+		originalShieldDamages[weaponDefID] = tonumber(weaponDef.customParams.shield_damage)
+	end
+
+	if weaponDef.type == 'Flame' then
+		flameWeapons[weaponDefID] = weaponDef
+	end
+end
+
+for id, data in pairs(UnitDefs) do
+	if data.customParams.shield_radius then
+		shieldUnitDefs[id] = data
+	end
+end
+
+function gadget:MetaUnitAdded(unitID, unitDefID, unitTeam)
+	if shieldUnitsData[unitID] then
+		shieldUnitsData[unitID].team = unitTeam
+	end
+end
+
+function gadget:UnitFinished(unitID, unitDefID, unitTeam)
+	if shieldUnitDefs[unitDefID] then
+		shieldUnitsData[unitID] = {
+			isStatic = shieldUnitDefs[unitDefID].isStatic or false,        -- Used to prefer paralyze disable method for shields that don't move
+			team = unitTeam,                                               -- For future AOE damage mitigation
+			location = { 0, 0, 0 },                                        -- For future AOE damage mitigation
+			shieldEnabled = true,                                          -- Virtualized enabled/disabled state until engine equivalent is changed
+			shieldDamage = 0,                                              -- This stores the value of damages populated in ShieldPreDamaged(), then applied in GameFrame() all at once
+			shieldWeaponNumber = -1,                                       -- This is replaced with the real shieldWeaponNumber as soon as the shield is damaged
+			downtime = shieldUnitDefs[unitDefID].customParams.shield_downtime or 8, -- Defined in unitdef.customparams with a default fallback value
+			downtimeReset = 0,
+			shieldCoverageChecked = false,									-- Used to prevent expensive unit coverage checks being performed more than once per cycle
+			radius = shieldUnitDefs[unitDefID].customParams.shield_radius
+		}
+	end
+
+	-- Increases performance by reducing global unitDefID lookups
+	unitDefIDCache[unitID] = unitDefID
+end
+
+function gadget:UnitDestroyed(unitID)
+	shieldUnitsData[unitID] = nil
+	unitDefIDCache[unitID] = nil
+end
+
+function gadget:ProjectileCreated(proID, proOwnerID, weaponDefID)
+	-- Increases performance by reducing global projectileDefID lookups
+	projectileDefIDCache[proID] = weaponDefID
+end
+
+function gadget:ProjectileDestroyed(proID)
+	projectileDefIDCache[proID] = nil
+	projectileShieldHitCache[proID] = nil
+end
```

Comment:
That's what the wiki says 🤷 
https://springrts.com/wiki/Lua_System#SetWatchWeapon

Reply:
In my unscientific test in game there was roughly 90% cache hit, so these callins are getting hit, but I think it's due to a side-effect of other gadgets registering Script.SetWatchProjectile.

Reply:
Clarification: if I needed to use the script.setwatchprojectile then I wouldn't be generating weapondefid's associated with projectiles. That would break the dependent parts of the gadget.

Reply:
`Script.SetWatchProjectile` (undocumented engine feature) takes a weapon def ID, and triggers the callins for projectiles created by that weapon def